pkg/webhook/server.go

```diff
@@ -203,105 +174,57 @@ func (s *Server) Start(stop <-chan struct{}) error {
+	cfg := &tls.Config{
+		Certificates: []tls.Certificate{cert},
 	}
 
-	shutdownHappend := false
-	timer := time.Tick(wait.Jitter(defaultCertRefreshInterval, 0.1))
-	go serveFn()
-	for {
-		select {
-		case <-timer:
-			changed, err := s.RefreshCert()
-			if err != nil {
-				log.Error(err, "encountering error when refreshing the certificate")
-				return err
-			}
-			if !changed {
-				log.Info("no need to reload the certificates.")
-				continue
-			}
-			log.Info("server is shutting down to reload the certificates.")
-			shutdownHappend = true
-			err = s.httpServer.Shutdown(context.Background())
-			if err != nil {
-				log.Error(err, "encountering error when shutting down")
-				return err
-			}
-			timer = time.Tick(wait.Jitter(defaultCertRefreshInterval, 0.1))
-			go serveFn()
-		case <-stop:
-			return s.httpServer.Shutdown(context.Background())
-		case e := <-errCh:
-			// Don't exit when getting an http.ErrServerClosed error due to restarting the server.
-			if shutdownHappend && e == http.ErrServerClosed {
-				shutdownHappend = false
-			} else if e != nil {
-				log.Error(e, "server returns an unexpected error")
-				return e
-			}
-		}
+	listener, err := tls.Listen("tcp", net.JoinHostPort("", strconv.Itoa(int(s.Port))), cfg)
+	if err != nil {
+		return err
 	}
-}
 
-// RefreshCert refreshes the certificate using Server's Provisioner if the certificate is expiring.
-func (s *Server) RefreshCert() (bool, error) {
-	cc, err := s.getClientConfig()
-	if err != nil {
-		return false, err
+	srv := &http.Server{
+		Handler: s.sMux,
 	}
-	changed, err := s.certProvisioner.Provision(cert.Options{
-		ClientConfig: cc,
-		Objects:      s.webhookConfigurations,
-	})
-	if err != nil {
-		return false, err
+
+	idleConnsClosed := make(chan struct{})
+	go func() {
+		<-stop
+
+		// TODO: use a context with reasonable timeout
+		if err := srv.Shutdown(context.Background()); err != nil {
+			// Error from closing listeners, or context timeout
+			log.Error(err, "error shutting down the HTTP server")
+		}
+		close(idleConnsClosed)
+	}()
+
+	err = srv.Serve(listener)
+	if err != nil && err != http.ErrServerClosed {
+		return err
 	}
 
-	return changed, batchCreateOrReplace(s.Client, s.webhookConfigurations...)
+	<-idleConnsClosed
+	return nil
 }
 
 var _ inject.Client = &Server{}
 
 // InjectClient injects the client into the server
 func (s *Server) InjectClient(c client.Client) error {
 	s.Client = c
-	for _, wh := range s.registry {
-		if _, err := inject.ClientInto(c, wh.Handler()); err != nil {
-			return err
-		}
-	}
-	return nil
-}
-
-var _ inject.Decoder = &Server{}
-
-// InjectDecoder injects the client into the server
-func (s *Server) InjectDecoder(d atypes.Decoder) error {
-	for _, wh := range s.registry {
-		if _, err := inject.DecoderInto(d, wh.Handler()); err != nil {
-			return err
+	for path := range s.registry {
+		// TODO(mengqiy): remove this in PR #316
+		if wh, ok := s.registry[path].(Webhook); ok {
```

Comment:
this inject code isn't quite right the way it's written, since someone could write their own webhook impl that needed the info.  If webhook isn't an interface, this isn't a problem.

Reply:
This is fixed in https://github.com/kubernetes-sigs/controller-runtime/pull/316/commits/a4ef92988bb52994fceff913d959f8e7f4162fa7#diff-fbc18bb07cdd05391b7081acc1dfe170R209
I tried to avoid putting everything in the same PR, so the code here may look incorrect.